### ISSUE TITLE
Add 'rot' argument to glyphInfo() to define rotation angle of shapes within a font

### ISSUE THREAD

**Issue by hturner** [r-devel/r-dev-day#18]
This issue to work on a feature request that has support from R Core developer Paul Murrell (@pmur002). 

It requires adapting both R and C code, as well as documentation, so could suit a team working on different aspects and then combining their patches.

The objective is to add a `rot` argument to `grDevices::glyphInfo()`, to define the rotation angle of glyphs (shapes within a text font). This is related to the new dev->glyph() API described in the report [Rendering Typeset Glyphs in R Graphics](https://www.stat.auckland.ac.nz/~paul/Reports/Typography/glyphs/glyphs.html).

TODO:

- [x] Add a `rot` argument to `grDevices::glyphInfo()` and update this R function so that the returned `"RGlyphInfo"` object includes the rotation angle.
- [x] Update the helpfile (glyphInfo.Rd) to match.
- [x] Update [`renderGlyphs`] C function to use the rotation angle from the glyphInfo argument. The functions that `renderGlyphs` calls (`getViewportTransform` and `GEGlyph`) are already designed to take a rotation angle, so it should just be a case of making sure the correct information is passed on.
- [x] Test the new functionality.
  - [x] include other than 90 degree multiple angles 
- [x] Move `glyph_rotation` after `glyph_colour`, so that using new `glyphInfo` works in old R, 
- [x] Add defensive code so that using old `glyphInfo` works with new R
- [x] Would ideally be nice to have demonstrations of those two scenarios working if possible:
  - [x] using new `glyphInfo` works in old R
  - [x] using old `glyphInfo` works with new R

**Comment by danielinteractive** [comment]
Thanks a lot @pmur002 for working with me on this exciting new feature!

I got a first prototype working 😄 

Quick question which would be good to confirm before proceeding to write the tests - I would propose the updated function signature:
```r
glyphInfo <- function(id, x, y, font, size,
                      fontList,
                      width, height,
                      hAnchor, vAnchor,
                      col=NA, rotation=0) { ... }
```

- This would ensure backwards compatibility, while using a readable argument name. 
- Alternative would be to stay with `rot` as described in the issue, which maybe important for consistency with other rotation arguments in `grid` and graphics packages? 
- It is not super consistent with `size` e.g. which occurs earlier and needs an explicit value, but I guess the alternative of putting `rotation` directly after `size` would not be backwards compatible.

Long story short: Is this signature ok or would you prefer an alternative one?

**Comment by danielinteractive** [comment]
```r
# Note for Windows users: 
#   This requires to use the AGG backend, 
#   select this in RStudio Graphics options and 
#   install the ragg package first.

library(grid)

# The following ids specify the 70th, 434th, 559th, etc glyphs within a font 
# (though not necessarily the same font):

ids <- c(70, 434, 559, 559, 391, 559, 475, 1642, 37, 475, 510, 521)

# The following locations describe a horizontal, left-to-right 
# placement of the glyphs (the x values increase monotonically 
# and the y values are constant):

x <- c(0, 7.5, 15.03125, 20.140625, 25.203125, 33.484375, 38.59375,
       42.1875, 45.421875, 55.328125, 58.671875, 66.84375)
y <- rep(0, 12)

# The following font indices and sizes specify that the first 
# seven glyphs all come from the same font (the first font in the
# fontList) and the remaining five glyphs all come from a different
# font (the second font in the fontList). The glyphs are all at the
# same font size (12pt).

fonts <- rep(1:2, c(7, 5))
size <- rep(12, 12)

# The following code describes two versions of the Google font 
# "Montserrat", one bold and italic and one plain. These fonts 
# are distributed as part of the 'grDevices' package:

font1 <- glyphFont(system.file("fonts", "Montserrat", "static",
                               "Montserrat-BoldItalic.ttf",
                               package="grDevices"),
                   0, "Montserrat", 700, "italic")
font2 <- glyphFont(system.file("fonts", "Montserrat", "static",
                               "Montserrat-Medium.ttf",
                               package="grDevices"),
                   0, "Montserrat", 400, "normal")

# The following width and height provide a bounding box for the 
# glyphs in our example:

width <- 74.46875
height <- 13.11719

# We specify that the first seven glyphs should be drawn in racing red:

rossoCorsa <- "#D40000"
col <- rep(c(rossoCorsa, NA), c(7, 5))

# Now we specify the rotation - first none:
rotation <- rep(0, length(size))

# The following code describes a set of twelve glyphs from two 
# Montserrat fonts, with the glyphs typeset horizontally from 
# left-to-right.

glyphs <- glyphInfo(ids, x, y, fonts, size,
                    glyphFontList(font1, font2),
                    width, height, col=col, rot=rotation)

# The following code draws the set of glyphs that we constructed
# above. We can now finally see that the glyphs that we have typeset
# spell out "Ferrari Dino".

grid.glyph(glyphs)

# OK so like this we can achieve the same behavior as before.

set.seed(123)
rotation2 <- runif(length(size), min = -360, max = 360)
glyphs2 <- glyphInfo(ids, x, y, fonts, size,
                    glyphFontList(font1, font2),
                    width, height, col=col, rot=rotation2)
grid.newpage()
grid.glyph(glyphs2)
# So it seems to work!

# We can also use a constant rotation across all letters:
grid.newpage()
grid.glyph(glyphInfo(ids, x, y, fonts, size,
                     glyphFontList(font1, font2),
                     width, height, col=col, rot=20))
```

**Comment by pmur002** [comment]
Thanks Daniel.  This looks excellent.  Two suggestions:

1.  It looks like the tests in `grid/tests/glyphs.R` all use rotations that are multiples of 90 degrees.  It would be nice to also cover some in-between angles.
2.  You have placed `glyph_rotation` ahead of the existing `glyph_colour` and therefore shifted `glyph_colour` down (`src/main/engine.c`).  This raises the issue of what to do if someone has saved a `glyphInfo` object from a previous R version and attempts to use it in the new R version.  The reverse is even possible, though less likely.  In the reverse case, adding `glyph_rotation` after `glyph_colour` (and leaving `glyph_colour` where it is) would still run.  For the case of encountering an old `glyphInfo` object, could we add some defensive code to check the number of columns in the `glyphInfo` object before we attempt to access `glyph_rotation`?

**Comment by danielinteractive** [comment]
Thanks a lot @pmur002 for your kind feeback. I will add some more different in-between angles for the tests. Just to confirm, for the `glyph_rotation` position, do you agree with the proposed position, plus adding defensive code which e.g. would return a rotation of 0 for old `glyphInfo` objects? Or would you prefer to move it after `glyph_colour`, plus adding defensive code?

**Comment by pmur002** [comment]
The latter I think.  Move `glyph_rotation` after `glyph_colour`, so that using new `glyphInfo` works in old R, plus add defensive code so that using old `glyphInfo` works with new R.  Would ideally be nice to have demonstrations of those two scenarios working if possible.

**Comment by danielinteractive** [comment]
Thanks again @pmur002 and apologies for the slight delay, I just moved across continents 😄 

Please see attached the updated patch: 
[glyphInfoRotation2.patch](https://github.com/user-attachments/files/16392404/glyphInfoRotation2.patch)
I successfully ran ` make -f Makefile.win test-Packages` with the new patch applied to current the R-devel. (`make check-devel` still fails for me with the unrelated `ERROR: this R is version 4.5.0, package 'pkgA' requires R >= r56550`)

- Rotation values which are not multiples of 90 degree are tested now
- Changed `rot` to be the last column in the data frame 
- The additional predicate function `R_GE_hasGlyphRotation` is used to be defensive before accessing the data frame column in `typeset.c`
- `drawDetails.glyphgrob()` is also defensive when determining which columns to use for calculating the runs of same graphics parameters

Here is an additional script which can be run with the patched R-devel as well as the previous R release (4.4.1) to demonstrate the backwards and forwards compatibility, using the `rds` files in 
[exampleGlyphInfo.zip](https://github.com/user-attachments/files/16392429/exampleGlyphInfo.zip) (unfortunately I needed to zip them together first to be able to upload it to GitHub here):

```r
library(grDevices)
library(grid)

testGlyphInfo <- list()
testGlyph <- function(info, ...) {
  testGlyphInfo[[length(testGlyphInfo) + 1]] <<- info
  if (!inherits(info, "RGlyphInfo")) {
    ## List of infos
    n <- length(info)
    for (i in 1:n) {
      vp <- viewport(y=i/(n+1))
      pushViewport(vp)
      grid.glyph(info[[i]], ...)
      popViewport()
    }
  } else {
    grid.glyph(info, ...)
  }
}
testGlyphGrob <- function(info, ...) {
  testGlyphInfo[[length(testGlyphInfo) + 1]] <<- info
  if (!inherits(info, "RGlyphInfo")) {
    ## List of infos
    n <- length(info)
    vps <- lapply(1:n, function(i) viewport(y=i/(n+1)))
    do.call(grobTree,
            mapply(function(x, vp) {
              glyphGrob(x, ..., vp=vp)
            },
            info, vps))
  } else {
    glyphGrob(info, ...)
  }
}

## glyphs
grid.newpage()
grid.segments(0,.5,1,.5, gp=gpar(col="grey"))

# Run in patched R devel:
oldGlyphInfo <- readRDS("oldGlyphInfo.rds")
testGlyph(oldGlyphInfo)

# Run in R release 4.4.1:
newGlyphInfo <- readRDS("newGlyphInfo.rds")
testGlyph(newGlyphInfo)
```

**Comment by pmur002** [comment]
Thanks Daniel.  Quick query:  are these lines in the patch necessary?  

```
+    ## Reorder to ensure backwards compatibility with code
+    ## where rot was not yet included.
+    col_order <- c("id", "x", "y", "font", "size", "colour", "rot")
+    glyphs <- glyphs[, col_order, drop = FALSE]
```

The code above those lines has explicitly created a data frame with columns in that order, so isn't this redundant?  Or am I missing something?

**Comment by pmur002** [comment]
Also checking: are these lines to allow for forward/backward compatibility (will handle old glyphInfo object as well as new)?

```
+    glyph_run_cols <- intersect(
+      c("font", "size", "rot", "colour"),
+      names(x$glyphInfo$glyphs)
+    )
```

**Comment by danielinteractive** [comment]
> Thanks Daniel. Quick query: are these lines in the patch necessary?
> 
> ```
> +    ## Reorder to ensure backwards compatibility with code
> +    ## where rot was not yet included.
> +    col_order <- c("id", "x", "y", "font", "size", "colour", "rot")
> +    glyphs <- glyphs[, col_order, drop = FALSE]
> ```
> 
> The code above those lines has explicitly created a data frame with columns in that order, so isn't this redundant? Or am I missing something?

Thanks @pmur002 , yeah we need this reordering because `rot` is treated together with the other non-`colour` columns, therefore sits before `colour` just before these 2 lines. It is not super efficient in terms of computation, but I felt that breaking the pattern in lines 93/94 by treating `rot` differently would be less preferable. Happy about changing that trade off, of course, if you think otherwise.

**Comment by danielinteractive** [comment]
> Also checking: are these lines to allow for forward/backward compatibility (will handle old glyphInfo object as well as new)?
> 
> ```
> +    glyph_run_cols <- intersect(
> +      c("font", "size", "rot", "colour"),
> +      names(x$glyphInfo$glyphs)
> +    )
> ```

Yes, exactly, this makes sure that if `rot` is not included in the `glyphs` data frame (i.e. old `glyphInfo`) that we don't include it when calculating the "same graph pars" runs (with new code).

**Comment by pmur002** [comment]
> Thanks @pmur002 , yeah we need this reordering because `rot` is treated together with the other non-`colour` columns, therefore sits before `colour` just before these 2 lines. It is not super efficient in terms of computation, but I felt that breaking the pattern in lines 93/94 by treating `rot` differently would be less preferable. Happy about changing that trade off, of course, if you think otherwise.

Got it now.  Thanks for the explanation.  Thanks also for confirming my other query.

Just need to get &*^)&**! `make check-all` to complete and then I will try a commit.

**Comment by pmur002** [comment]
Committed to r-devel (r86937).  Thanks again!

**Comment by danielinteractive** [comment]
Thanks a lot @pmur002 ! That is exciting!
